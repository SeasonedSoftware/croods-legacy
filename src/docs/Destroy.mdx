---
name: Destroy
route: /destroy
---

import { Playground } from 'docz'

# Destroy

The Destroy component provides an easy way to destroy a resource in the server, using the response as needed.

## API

| Property | Type          | Required | Default          | Description                                                                                                                          |
| -------- | ------------- | -------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| name     | String        | true     | -                | It must be equal to the reducer name set in your `rootReducer.js` and will be used as `path` if none is provided Ex: `baseUrl/name`. |
| id       | String/Number | true     | -                | It defines the ID of the resource to be destroyed, passing it in the url as defined in path.                                         |
| render   | Func          | true     | -                | A function returning a React Node. Ex: ({ list }, props) -> < JSX />                                                                 |
| parentId | String/Number | false    | -                | Parent element's ID, for nested resources.                                                                                           |
| path     | String        | false    | `/${name}/${id}` | Ex: /foo/123                                                                                                                         |

## How to use

Don't forget to add a reducer with the same name used here in your rootReducer, with **createReducer**

```javascript
import { Destroy } from 'croods'

export default (
  <Destroy
    name="pets"
    id="123"
    // if path is different than /pets/${id}, you can set a custom path like so:
    // path='/customPath'
    render={destroy => <DestroyButton destroy={destroy} />}
  />
)
```

### render

The render function is used to render a component passing to it a **destroy** function, and
pass it to any component you like.

## Example

You can see a working example using this component in [here](https://github.com/SeasonedSoftware/croods/blob/master/example/src/colors/Destroy.js).
