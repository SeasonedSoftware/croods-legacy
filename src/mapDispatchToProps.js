import { bindActionCreators } from 'redux'

import createActions from './createActions'

export default (dispatch, { options, ...props }) => ({
  actions: bindActionCreators(
    createActions({ ...options, ...props }),
    dispatch,
  ),
})
