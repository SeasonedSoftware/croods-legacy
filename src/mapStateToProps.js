import find from 'lodash/find'
import get from 'lodash/get'

export default (state, { name, parentId }) => {
  const statePart = get(state, name)
  const currentState = find(statePart, block => block.parentId === parentId)
  return get(currentState, 'state') || {}
}
