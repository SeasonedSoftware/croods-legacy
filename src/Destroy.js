import { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import customPropTypes from './customPropTypes'
import providerProps from './providerProps'
import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import resetIfChanged from './resetIfChanged'

class Destroy extends Component {
  constructor(props) {
    super(props)
    resetIfChanged({ props, prevProps: {}, name: 'destroyed' })
  }

  componentDidUpdate(prevProps) {
    resetIfChanged({ props: this.props, prevProps, name: 'destroyed' })
  }

  render() {
    const { render, id, actions } = this.props
    const destroy = () => actions.destroy(id)

    return render(destroy, this.props)
  }
}

Destroy.propTypes = {
  id: customPropTypes.id.isRequired,
  name: customPropTypes.name.isRequired,
  render: PropTypes.func.isRequired, // (destroy, props) -> Html
  parentId: customPropTypes.id,
  path: customPropTypes.path,
  // parse responses to adjust the API to croods patterns
  parseResponse: PropTypes.func, // (json, response, requestAttributes) -> Object

  ...providerProps,
}

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Destroy),
)
