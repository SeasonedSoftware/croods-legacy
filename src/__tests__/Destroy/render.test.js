import React from 'react'
import { connect } from 'react-redux'
import renderer from 'react-test-renderer'

import Destroy from '../../Destroy'
import mapStateToProps from '../../mapStateToProps'
import mapDispatchToProps from '../../mapDispatchToProps'

jest.mock('../../withOptions', () => Component => Component)

jest.mock('react-redux', () => ({
  connect: jest.fn(() => Component => Component),
}))

const render = jest.fn((destroy, props) => {
  destroy && destroy()
  return (
    <div destroy={destroy} {...props}>
      Destroy
    </div>
  )
})

const actions = { destroy: jest.fn(), resetDestroyed: jest.fn() }

beforeEach(() => {
  render.mockClear()
  actions.resetDestroyed.mockClear()
})

it('renders correctly', () => {
  const props = { render, actions, id: 123 }
  const tree = renderer.create(<Destroy {...props} />).toJSON()

  expect(tree).toMatchSnapshot()
  expect(connect).toHaveBeenCalledWith(mapStateToProps, mapDispatchToProps)
  expect(actions.resetDestroyed).not.toHaveBeenCalled()
  expect(actions.destroy).toHaveBeenCalledWith(123)
})

describe('when destroyed', () => {
  it('resets destroyed', () => {
    const destroyed = { id: 123 }
    const props = { render, actions, id: 123, destroyed }
    const tree = renderer.create(<Destroy {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(connect).toHaveBeenCalledWith(mapStateToProps, mapDispatchToProps)
    expect(actions.resetDestroyed).toHaveBeenCalledWith()
  })
})
