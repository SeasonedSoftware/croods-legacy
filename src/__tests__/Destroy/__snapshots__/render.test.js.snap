// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  actions={
    Object {
      "destroy": [Function],
      "resetDestroyed": [Function],
    }
  }
  destroy={[Function]}
  id={123}
  render={[Function]}
>
  Destroy
</div>
`;

exports[`when destroyed resets destroyed 1`] = `
<div
  actions={
    Object {
      "destroy": [Function],
      "resetDestroyed": [Function],
    }
  }
  destroy={[Function]}
  destroyed={
    Object {
      "id": 123,
    }
  }
  id={123}
  render={[Function]}
>
  Destroy
</div>
`;
